src/services/gcs/backend.rs

```diff
@@ -0,0 +1,681 @@
+            debug!("object {} create finished", args.path());
+            Ok(())
+        } else {
+            warn!(
+                "object {} create returned status code: {}",
+                args.path(),
+                resp.status()
+            );
+            let er = parse_error_response(resp).await?;
+            let e = parse_error("create", args.path(), er);
+            Err(e)
+        }
+    }
+    async fn read(&self, args: &OpRead) -> Result<BytesReader> {
+        let p = self.get_abs_path(args.path());
+        debug!(
+            "object {} read start: offset {:?}, size {:?}",
+            &p,
+            args.offset(),
+            args.size()
+        );
+
+        let resp = self
+            .get_object(&p, args.offset(), args.size())
+            .await
+            .map_err(|e| {
+                error!("object {} get_object: {:?}", p, e);
+                e
+            })?;
+
+        if resp.status().is_success() {
+            debug!(
+                "object {} reader created: offset {:?}, size {:?}",
+                &p,
+                args.offset(),
+                args.size()
+            );
+
+            Ok(Box::new(resp.into_body()))
+        } else {
+            warn!(
+                "object {} read with status code: {}",
+                args.path(),
+                resp.status()
+            );
+            let er = parse_error_response(resp).await?;
+            let e = parse_error("read", args.path(), er);
+            Err(e)
+        }
+    }
+
+    async fn write(&self, args: &OpWrite) -> Result<BytesWriter> {
+        let p = self.get_abs_path(args.path());
+        debug!("object {} write start: size {}", &p, args.size());
+
+        let (tx, body) = new_http_channel(args.size());
+
+        let req = self.insert_object(&p, args.size(), body).await?;
+
+        let bs = HttpBodyWriter::new(
+            args,
+            tx,
+            self.client.send_async(req),
+            |c| (200..300).contains(&c.as_u16()),
+            parse_error,
+        );
+
+        Ok(Box::new(bs))
+    }
+
+    async fn stat(&self, args: &OpStat) -> Result<ObjectMetadata> {
+        let p = self.get_abs_path(args.path());
+        debug!("object {} stat start", &p);
+
+        // Stat root always returns a DIR.
+        if self.get_rel_path(&p).is_empty() {
+            let mut m = ObjectMetadata::default();
+            m.set_mode(ObjectMode::DIR);
+
+            debug!("backed root object stat finished");
+            return Ok(m);
+        }
+
+        let mut resp = self.get_object_metadata(&p).await?;
+
+        if resp.status().is_success() {
+            let mut m = ObjectMetadata::default();
+            // read http response body
+            let mut v = Vec::new();
+            resp.body_mut().read(&mut v).await.map_err(|e| {
+                error!("GCS backend failed to read response body: {:?}", e);
+                e
+            })?;
+            let j_object: Value =
```

Comment:
I don't like the idea of using `serde_json::Value`.

Please declare a struct and derive `serde::Deserialize` for it. So we can use:

```let
let gcs_meta = de::from_slice(&bs).map_err(xxxx);
```

Simple and beautiful.